linkerd/http-route/src/http/filter/redirect.rs

```diff
@@ -0,0 +1,416 @@
+        match (self.host.as_deref(), self.port) {
+            // If a host is configured, use it and whatever port is configured.
+            (Some(h), Some(p)) => format!("{}:{}", h, p).parse().map_err(Into::into),
+            (Some(h), None) => h.parse().map_err(Into::into),
+
+            // If a host is NOT configured, use the request's original host
+            // and either an overridden port or the original port.
+            (None, p) => {
+                let h = orig_uri.host().ok_or(InvalidRedirect::MissingAuthority)?;
+                match p.or_else(|| orig_uri.port_u16().and_then(|p| p.try_into().ok())) {
+                    Some(p) => format!("{}:{}", h, p).parse().map_err(Into::into),
+                    None => h.parse().map_err(Into::into),
+                }
+            }
+        }
+    }
+
+    // XXX This function probably does more allocation that is strictly needed.
+    // We may want to optimize it as it settles.
+    fn path_and_query(
+        &self,
+        orig_uri: &http::Uri,
+        rm: &RouteMatch,
+    ) -> Result<PathAndQuery, InvalidRedirect> {
+        use crate::http::r#match::PathMatch;
+
+        match &self.path {
+            // If the redirect does not specify a path, use the original path/query.
+            None => Ok(orig_uri
+                .path_and_query()
+                .expect("URI must have a path")
+                .clone()),
+
+            // If the redirect specifies a full path (potentially including a
+            // query), use it.
+            Some(ModifyPath::ReplaceFullPath(p)) => p.clone().try_into().map_err(Into::into),
+
+            // If the redirect specifies a prefix rewrite, using the original
+            // query parameters.
+            //
+            // XXX #fragments are not included in the rewritten location; but
+            // fragments are generally not transmitted to servers.
+            Some(ModifyPath::ReplacePrefixMatch(new_pfx)) => match rm.route.path() {
+                PathMatch::Prefix(pfx_len) if *pfx_len <= orig_uri.path().len() => {
+                    let mut new_path = new_pfx.to_string();
+                    let (_, rest) = orig_uri.path().split_at(*pfx_len);
+                    if !rest.is_empty() && !rest.starts_with('/') {
+                        new_path.push('/');
+                    }
+                    new_path.push_str(rest);
+                    if let Some(q) = orig_uri.query() {
+                        new_path.push('?');
+                        new_path.push_str(q);
+                    }
+                    new_path.try_into().map_err(Into::into)
+                }
+
+                // If the matched rule was not a prefix match, the redirect
+                // filter is invalid. This should cause us to fail requests with
+                // a 5XX.
+                _ => Err(InvalidRedirect::InvalidReplacePrefix),
+            },
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
thanks for adding tests for this!

as a side note, this feels like the kind of code that _might_ be well-suited by property testing...but i'm not sure if it's worth the effort to get `quickcheck`/`proptest` to generate arbitrary valid URIs or not...